server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -561,35 +582,33 @@ private static void parseArray(ParseContext context, ObjectMapper parentMapper,
     }
 
     private static void parseNonDynamicArray(ParseContext context, ObjectMapper mapper,
-                                             String lastFieldName, String arrayFieldName) throws IOException {
+                                             final String lastFieldName, String arrayFieldName) throws IOException {
         XContentParser parser = context.parser();
         XContentParser.Token token;
+        final String[] paths = splitAndValidatePath(lastFieldName);
         while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
             if (token == XContentParser.Token.START_OBJECT) {
-                parseObject(context, mapper, lastFieldName);
+                parseObject(context, mapper, lastFieldName, paths);
             } else if (token == XContentParser.Token.START_ARRAY) {
-                parseArray(context, mapper, lastFieldName);
-            } else if (token == XContentParser.Token.FIELD_NAME) {
```

Comment:
This statement is impossible to reach, there shouldn't be a field name as a direct child of an array ever? Removed it so the logic for keeping the paths array is a little simpler to follow.